### PR TITLE
Issue2024 ashrae2006 sequence

### LINKED ISSUE #2024: ASHRAE2006: Economizer controlled in sequence with the heating and cooling coils
Currently in `Buildings.Examples.VAVReheat.ASHRAE2006` the control of the economizer is based on `TMix` and the control of the heating and cooling coils is based on `TSup` (with 2 independent control loops). These 3 different control loops use 3 different set points, namely 10 degC for the heating coil, 12 degC for the cooling coil and the average (11 degC) for the economizer.
This is not as described in the documentation of the control sequence 2A2-21232 which is cited as a reference in the model information section.

> The heating coil valve, mixed air dampers, and cooling coil valve shall modulate in sequence to maintain supply air temperature setpoint of 55°F (adjustable). Provide deadband between heating and economizer cooling.

This issue is to implement this exact control logic, using one single controller and a mapping function to sequence the 3 control actions.

[EDIT 1] During the implementation comes the question of set point scheduling. The current model in the master uses

- a constant supply set point for the heating coil (10 degC), independent from the operation mode,
- a scheduled supply set point for the cooling coil, namely
    - `max(TSetHea+1, TCooOn)` with `TCooOn=12` degC in occupied, unoccupiedPreCool and safety modes
    - `TCooOff=30` degC in unoccupiedOff, unoccupiedNightSetBack and unoccupiedWarmUp modes

This cannot be used anymore when implementing the control sequence 2A2-21232 since the heating coil and cooling coil share the same supply temperature set point (so using `TCooOff` would increase the load on the heating coil in unoccupied mode). Therefore, the set point scheduling for the cooling coil will not be used anymore. No impact is expected as 

- The fan is off in unoccupiedOff mode.
- The unoccupiedNightSetBack and unoccupiedWarmUp modes are activated based on a test for a heating demand in the rooms. So no load on the cooling coil is expected in this case.

[EDIT 2] There is another issue in the master related to the operation modes and corresponding set points. Currently:

- In `Buildings.Examples.VAVReheat.Controls.Economizer` `connect(max.y, extractor.u[Integer(OperationModes.unoccupiedWarmUp)])` whereas the ASHRAE VAV 2A2-21232 states
  > Morning Warm-up: The optimum start program shall start the unit at the latest possible
time to reach the desired occupied space temperature setpoint at occupancy time. If the
average space temperature is below the occupied space temperature setpoint, the supply
and return fans shall energize, the maximum and minimum outside and exhaust dampers
shall remain closed, and return damper open

- In `Buildings.Examples.VAVReheat.Controls.ModeSelector` the condition `TRooAve.y < TRooSetCooOcc` is used to transition towards warm-up mode whereas the ASHRAE VAV 2A2-21232 states
  > If the average space temperature is below the occupied space temperature setpoint...

- In `Buildings.Examples.VAVReheat.ASHRAE2006` the heating coil air outlet temperature set point is constant (10 degC) whereas the ASHRAE VAV 2A2-21232 states
  > The heating coil valve shall modulate to maintain supply air temperature setpoint of 95°F (adjustable) until the average space temperature equals the occupied space temperature setpoint. 

  The current setting where the set point in warm-up is lower than the room set point (and thus the return temperature), raises an issue when transitioning to a control in sequence of the supply temperature. In warm-up mode (with full air recirculation) the controller yields a cooling demand but gets "incorrect feedback" from the system as the economizer is disabled. So when the mode changes to "occupied" the economizer starts will fully open OA dampers (due to the saturated control signal yielded by the supply temperature controller) and the time it takes for the control output to desaturate and activate the heating coil, freezing conditions are reached in the cooling coil.

### PR DESCRIPTION
**Pull request by AntoineGautier**
This closes #2024.

### CODE REVIEW

**Comment by AntoineGautier** [comment]
This PR is converted into draft as other models depend on the refactoring of the economizer sequence and a decision has to be made before propagating refactoring these models.
`Buildings.Examples.DualFanDualDuct.ClosedLoop`
`Buildings.Examples.ScalableBenchmarks.BuildingVAV.Examples.OneFloor_OneZone`

I would suggest to add a boolean parameter to `Buildings.Examples.VAVReheat.Controls.SupplyAirTemperature` so that it can be used for cooling only (for `DualFanDualDuct`), or heating and cooling (for single duct VAV).
Waiting for @mwetter's approval to move forward.

The impact of the control sequence refactoring has been assessed on commit 5d537ab, it is not significant.

**Branch `issue2024_ASHRAE2006_sequence`**

```
. table:: Heating, cooling, fan and total site HVAC energy, and savings of guideline 36 case versus base case.

   ===================================== ===================================== ====================================== ====================================== =====
   :math:`E_{h} \quad [kWh/(m^2\,a)]`    :math:`E_{c} \quad [kWh/(m^2\,a)]`    :math:`E_{f} \quad [kWh/(m^2\,a)]`     :math:`E_{tot} \quad [kWh/(m^2\,a)]`     [%]
   ===================================== ===================================== ====================================== ====================================== =====
                                   10.17                                 19.71                                  2.384                                  32.26   
                                   2.961                                 14.83                                  1.868                                  19.66  39.1 
   ===================================== ===================================== ====================================== ====================================== =====
  
  
CPUtime, base 54'
CPUtime, G36  94'
```

**Branch `master`**

```
.. table:: Heating, cooling, fan and total site HVAC energy, and savings of guideline 36 case versus base case.

   ===================================== ===================================== ====================================== ====================================== =====
   :math:`E_{h} \quad [kWh/(m^2\,a)]`    :math:`E_{c} \quad [kWh/(m^2\,a)]`    :math:`E_{f} \quad [kWh/(m^2\,a)]`     :math:`E_{tot} \quad [kWh/(m^2\,a)]`     [%]
   ===================================== ===================================== ====================================== ====================================== =====
                                   10.15                                 19.89                                  2.389                                  32.42   
                                   2.961                                 14.83                                  1.868                                  19.66  39.4 
   ===================================== ===================================== ====================================== ====================================== =====
  
  
CPUtime, base 52'
CPUtime, G36  94'
```

**Comment by AntoineGautier** [comment]
TODO

- [ ] Check that a dead band is not needed between economizer cooling and coil cooling: is the economizer control stable or does it oscillates with coil modulating?
- [ ] Add a boolean parameter to the sequence for heating/cooling only mode needed for dual duct.
- [ ] The minimum TMix control by the economizer is not mentioned in the control sequence VAV 2A2-21232 of the Sequences of Operation for Common HVAC Systems (the heating coil ensures the preheat function). Adapt the sequence to eliminate this control loop. Verify that a constant flow pump ("freeze pump") is not needed for the heating coil then and adapt the heating coil control consistently.

**Comment by mwetter** [comment]
@AntoineGautier : Adding a boolean parameter to `Buildings.Examples.VAVReheat.Controls.SupplyAirTemperature` so that it can be used for cooling only (for `DualFanDualDuct`) is fine.

**Comment by AntoineGautier** [comment]
With commit b9bd3cd the simulation of the models in `Examples` hits a timeout with OCT and JModelica on Travis.
However, I cannot reproduce locally.

ter

When comparing the branch to the master I get the following statistics with OCT (on the models with the highest simulation time).

oct

There is definitely an increase in simulation time for some models (such as the ones in `ScalableBenchmarks.BuildingVAV.Examples`) due to an increase in the number of state events. 
This is due to the correction of the supply air temperature set point in Unoccupied Setback mode, see EDIT 2 in #2024.
In the master the setback has no effect due to the low supply temperature, see plot below.
With the correction, the mode switches between setback and unoccupied off as it should be.

`master`

master

`issue2024_ASHRAE2006_sequence`

dev

**Comment by AntoineGautier** [comment]
@JayHuLBL This is ready for your review when all tests pass.

- [ ] Could you explain in the documentation of `Buildings.Examples.VAVReheat.Controls.Economizer` the choice for the default values of `yMax = 0.995` and `yMax = 0.005`?

On my end, I will run nightly simulations on whole year and 

- [ ] update the impact on the energy use comparison between ASHRAE2006 and Guideline 36,
- [ ] check that a dead band is not needed between economizer cooling and coil cooling (is the economizer control stable or does it oscillates with coil modulating?).

File: Buildings/Examples/VAVReheat/Controls/Economizer.mo
Comment (JayHuLBL) [inline Buildings/Examples/VAVReheat/Controls/Economizer.mo:84]: Correct the comment `Takes smaller signal (OA damper opens for freeze protection control or for minimum outside air.)`

**Comment by JayHuLBL** [comment]
@AntoineGautier In instance `conV_flow` of class `Buildings.Examples.VAVReheat.Controls.Economizer`, by setting `yMin=0.005` may allow freeze protection control output being selected when it comes to the lower end (if `have_frePro=true`), or avoid fully closed damper (if `have_frePro=false`); For `yMax=0.995`, I also don't understand.

File: Buildings/Examples/VAVReheat/Controls/Economizer.mo
Comment (JayHuLBL) [inline Buildings/Examples/VAVReheat/Controls/Economizer.mo:175]: add default component name

File: Buildings/package.mo
Comment (JayHuLBL) [inline Buildings/package.mo:322]: remove the empty space at the end

File: Buildings/package.mo
Comment (JayHuLBL) [inline Buildings/package.mo:497]: empty space

**Review by JayHuLBL — COMMENTED** [review]
@AntoineGautier See the inline comments. Otherwise, it's good to merge.

File: Buildings/Examples/VAVReheat/Controls/Economizer.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/Economizer.mo:84]: Addressed

File: Buildings/Examples/VAVReheat/Controls/Economizer.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/Economizer.mo:175]: Addressed

File: Buildings/package.mo
Comment (AntoineGautier) [reply Buildings/package.mo:322]: Addressed (no idea how it got there in the first place)

File: Buildings/package.mo
Comment (AntoineGautier) [reply Buildings/package.mo:497]: Addressed

**Comment by AntoineGautier** [comment]
@milicag You probably have a derived class of `Examples.VAVReheat.ASHRAE2006` using SOEP room model in `issue1502_new_add_of_replaceable_thermal_zone_models`. Note that this PR will change the simulation results of such a model. It should not cause any simulation issue though but please double check.

**Comment by AntoineGautier** [comment]
@milicag Can I go ahead and merge this?

**Comment by milicag** [comment]
@AntoineGautier it looks like I deleted one of your comments as I thought it is my reply to the comment that I am deleting, since I found the answer already. Sorry about that.

File: Buildings/Examples/VAVReheat/Controls/Examples/SupplyAirTemperature.mo
Comment (milicag) [inline Buildings/Examples/VAVReheat/Controls/Examples/SupplyAirTemperature.mo:14]: I think we're using setpoint throughout the library.

File: Buildings/Examples/VAVReheat/Controls/ModeSelector.mo
Comment (milicag) [inline Buildings/Examples/VAVReheat/Controls/ModeSelector.mo:106]: Maybe better named `yEcoSta`?

File: Buildings/Examples/VAVReheat/Controls/SupplyAirTemperatureSetpoint.mo
Comment (milicag) [inline Buildings/Examples/VAVReheat/Controls/SupplyAirTemperatureSetpoint.mo:1]: Again I see a few usages `set point`. While I do not mind or have preference, I do think this should be harmonized within the library.

**Review by milicag — CHANGES_REQUESTED** [review]
@AntoineGautier I have a few minor comments. Other than those it looks fine to me.

File: Buildings/Examples/VAVReheat/Controls/SupplyAirTemperatureSetpoint.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/SupplyAirTemperatureSetpoint.mo:1]: This will be addressed through #2225

File: Buildings/Examples/VAVReheat/Controls/ModeSelector.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/ModeSelector.mo:106]: I would prefer not to use status for a command signal (digital output) and reserve its use for the return signal from the equipment (digital input). This is consistent with Guideline 36 and RP1711 I believe.

File: Buildings/Examples/DualFanDualDuct/ClosedLoop.mo
Comment (AntoineGautier) [reply Buildings/Examples/DualFanDualDuct/ClosedLoop.mo:358]: I don't see the reason why to make this parameter binding final here. The user should be able to modify these values if needed.

File: Buildings/Examples/VAVReheat/Controls/Economizer.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/Economizer.mo:12]: Addressed

File: Buildings/Examples/VAVReheat/Controls/Examples/SupplyAirTemperature.mo
Comment (AntoineGautier) [reply Buildings/Examples/VAVReheat/Controls/Examples/SupplyAirTemperature.mo:14]: This will be addressed through #2225

**Comment by AntoineGautier** [comment]
@mwetter This is ready to merge when all tests pass.

**Review by milicag — APPROVED** [review]

**Pull request merged**
